File: client/nodeid.go
// SPDX-FileCopyrightText: Ella Networks Inc.
// SPDX-License-Identifier: BUSL-1.1

package client

import (
	"encoding/json"
	"strconv"
	"strings"
)

const (
	minLegacyNodeID = 1
	maxLegacyNodeID = 63
)

type NodeID string

func (n NodeID) String() string {
	return string(n)
}

func (n NodeID) MarshalJSON() ([]byte, error) {
	if isDecimal(string(n)) {
		return []byte(n), nil
	}

	return json.Marshal(string(n))
}

func (n *NodeID) UnmarshalJSON(data []byte) error {
	raw := strings.TrimSpace(string(data))
	if raw == "null" {
		*n = ""

		return nil
	}

	if strings.HasPrefix(raw, `"`) {
		var s string
		if err := json.Unmarshal(data, &s); err != nil {
			return err
		}

		raw = s
	}

	*n = NodeID(strings.ToLower(strings.TrimSpace(raw)))

	return nil
}

func isDecimal(s string) bool {
	n, err := strconv.ParseUint(s, 10, 31)
	if err != nil {
		return false
	}

	if strconv.FormatUint(n, 10) != s {
		return false
	}

	return n >= minLegacyNodeID && n <= maxLegacyNodeID
}

File: ui/src/queries/nodeId.ts
// SPDX-FileCopyrightText: Ella Networks Inc.
// SPDX-License-Identifier: BUSL-1.1
export type NodeId = string | number;

export const nodeIdKey = (id: NodeId | undefined): string =>
  id === undefined ? "" : String(id);

export const sameNodeId = (
  a: NodeId | undefined,
  b: NodeId | undefined,
): boolean => a !== undefined && b !== undefined && String(a) === String(b);

export const shortNodeId = (id: NodeId | undefined): string => {
  const s = nodeIdKey(id);

  return s.length > 12 ? s.slice(-12) : s;
};

export const nodeLabel = (
  displayName: string | undefined,
  id: NodeId | undefined,
): string =>
  displayName && displayName.length > 0 ? displayName : shortNodeId(id);
